### ISSUE TITLE
[web]: Not Able to Receive SSE Events

### ISSUE THREAD

**Issue by ponderMuse** [flutter/flutter#43343]
Have a Flutter App that is able to receive SSE text/event-stream events when running on mobile but it is not receiving the same SSE text/event-stream events when running on the chrome web.

	import 'package:flutter/material.dart';
	import 'package:http/http.dart' as http;

	void main() => runApp(MyApp());

	class MyApp extends StatelessWidget {

	  http.Client _client;

	  MyApp() : super() {
		print("Ctor called");
		subscribe();
	  }

	  @override
	  Widget build(BuildContext context) {
		print("building..");
		return MaterialApp(
		  title: 'Flutter SSE',
		  home: Scaffold(
			appBar: AppBar(
			  title: Text('Receive SSE Events'),
			),
			body: Center(
			  child: Text('Ready for events..'),
			),
		  ),
		);
	  }

	  subscribe() async {
		print("Subscribing..");
		try {
		  _client = http.Client();

		  var request = new http.Request("GET", Uri.parse("http://192.168.1.11:8080/myserver/events"));
		  request.headers["Cache-Control"] = "no-cache";
		  request.headers["Accept"] = "text/event-stream";

		  Future<http.StreamedResponse> response = _client.send(request);
		  print("Subscribed!");

		  response.asStream().listen((streamedResponse) {
			print("Received streamedResponse.statusCode:${streamedResponse.statusCode}");
			streamedResponse.stream.listen((data) {
			  print("Received data:$data");
			});
		  });
		} catch(e) {
		  print("Caught $e");
		}
	  }

	  unsubscribe() {
		_client.close();
	  }
	}

On the server-side, I can see that the flutter app always subscribes okay as I can see the Sink being added. And the server-side always dispatches the events to the Sink but the flutter app only receives the events when running as a mobile app and NOT when running as a Chrome web app.

C:\Users\pondeMuse\flutter\bin\flutter.bat doctor --verbose
[√] Flutter (Channel master, v1.10.15-pre.188, on Microsoft Windows [Version 10.0.16299.1087], locale en-GB)
    • Flutter version 1.10.15-pre.188 at C:\Users\pondeMuse\flutter
    • Framework revision 584ee10c68 (2 days ago), 2019-10-21 07:49:28 -0700
    • Engine revision 8aefcd8575
    • Dart version 2.6.0 (build 2.6.0-dev.8.0 a61c775db8)

[√] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
    • Android SDK at C:\Users\pondeMuse\AppData\Local\Android\sdk
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-29, build-tools 29.0.2
    • Java binary at: C:\Program Files\Android\Android Studio\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 1.8.0_202-release-1483-b03)
    • All Android licenses accepted.

[√] Chrome - develop for the web
    • Chrome at C:\Program Files (x86)\Google\Chrome\Application\chrome.exe

[√] Android Studio (version 3.5)
    • Android Studio at C:\Program Files\Android\Android Studio
    • Flutter plugin version 40.2.2
    • Dart plugin version 191.8593
    • Java version OpenJDK Runtime Environment (build 1.8.0_202-release-1483-b03)

[!] IntelliJ IDEA Community Edition (version 2019.2)
    • IntelliJ at C:\Program Files\JetBrains\IntelliJ IDEA Community Edition 2018.2.5
    X Flutter plugin not installed; this adds Flutter specific functionality.
    • Dart plugin version 192.7402
    • For information about installing plugins, see
      https://flutter.dev/intellij-setup/#installing-the-plugins

[√] Connected device (2 available)
    • Chrome     • chrome     • web-javascript • Google Chrome 77.0.3865.120
    • Web Server • web-server • web-javascript • Flutter Tools

! Doctor found issues in 1 category.

**Comment by ponderMuse** [comment]
I keep reading that there could be issues associated with CORS when running on a web browser but as far as I can see, there are no issues with CORS. CORS on the server-side has been configured to accept all origins (i.e. "*"), configured to accept "GET" and "OPTIONS" methods, and configured to accept all headers (i.e. "*").

Debugging in Chrome browser shows that the response back is a 200 and can see the following response headers:

	Access-Control-Allow-Origin: *
	Cache-Control: no-cache, no-store, max-age=0, must-revalidate
	Content-Type: text/event-stream;charset=UTF-8
	Expires: 0
	Pragma: no-cache
	transfer-encoding: chunked
	Vary: Origin
	Vary: Access-Control-Request-Method
	Vary: Access-Control-Request-Headers
	X-Content-Type-Options: nosniff
	X-Frame-Options: DENY
	X-XSS-Protection: 1 ; mode=block

Unfortunately, I don't know enough about the http protocol so I don't know if any of the response headers above should be reason for concern as to why an event stream cannot be received on a web browser?

As a side note, the flutter/dart code was originally written as angulardart/dart code and that code was able to receive streamed events fine when running on Chrome web browser (Therefore assuing that CORS was configured correctly).

Any help or pointers on this issue would be greatly appreciated. Even a simple "Yes" receiving streamed events is possible in current version of flutter web on Chrome so you must be doing something wrong on the server side. Or "No" this does not work in current version of flutter web on Chrome but eventually it will/never will, etc.

**Comment by yjbanov** [comment]
@vsmenon does this look like a Dart issue? There's not much we can do on the Flutter side. This is a network issue, not a UI issue.

**Comment by vsmenon** [comment]
@natebosch @grouma - any ideas / tips?

**Comment by ponderMuse** [comment]
Just for info, the backend server which generates the event stream is a Spring Boot 2 application using spring-webflux for the reactive streams and the app uses spring-security too. I'd like to think that the problem is somewhere in the backend in terms of the whole Spring configuration but I can't see any errors or anything like that anywhere (backend or frontend in Chrome browser). The response from the call to the event stream API is a 200 and there are no CORS related errors showing up (see my previous comment) or anything like that so I don't know where to look next.

One thing I should add here also is that, originally, the client web app was written using Angulardart and at that time, the web app on a Chrome browser was able to receive the event streams without issue. But now I have rerun the Angulardart client on Chrome and now the Angulardart app is also NOT receiving the event stream just like the flutter web app running on Chrome.

I should add here that, both the Angulardart client and Flutter web client share most of the same code as it is mostly all in a pure dart library with only the Widgets for flutter and Components for Angulardart being on their own projects. So it is important to note here that the dependencies' versions that the Angulardart used from the time it worked to the time now when it no longer works have indeed been updated. But unfortunately the dependency version changes have not been recorded in git commits or anything like that just yet so I can't say specifically what libraries have been updated between then and now (But they would definitely include updates to http library)

**Comment by ponderMuse** [comment]
I have written a simple nodejs server to test with minimal code:
	var http = require("http");

	http.createServer(function (req, res) {
	  var fileName = "." + req.url;

	  if (fileName === "./myserver/events") {
		console.log("Responding to request with events")
		res.writeHead(200, {
		  "Content-Type":"text/event-stream",
		  "Connection":"keep-alive",
		  "Access-Control-Allow-Origin": '*',
		  "Access-Control-Allow-Headers": '*'
		});
		res.write("retry: 10000\n");
		res.write("event: connecttime\n");
		res.write("data: " + (new Date()) + "\n\n");
		res.write("data: " + (new Date()) + "\n\n");

		interval = setInterval(function () {
			console.log("Sending event..")
		  res.write("data: " + (new Date()) + "\n\n");
		}, 1000);

		req.connection.addListener("close", function () {
		  clearInterval(interval);
		}, false);
	  }
	}).listen(8080, "192.168.1.11");

And it shows the same behaviour. Running the flutter app an an emulated mobile app the events are received okay. But running the flutter app on Chrome web the events are sent by server but not received by Chrome web app. Chrome debugging shows the network time lapse updating every time the server sends an event but the events themselves are never logged in the console or on the App's Widget like it does when run as a mobile app.

**Comment by grouma** [comment]
Can you reproduce this issue without Flutter Web and strictly `[package:sse](https://github.com/dart-lang/sse)`? If so let's open a tracking issue on that package to help get to the bottom of this.

**Comment by ponderMuse** [comment]
@grouma, how should I get the sse client running on a chrome browser if not using Flutter Web? Should I write it using Angluardart?

Using strictly [package:sse](https://github.com/dart-lang/sse) and nothing else then the client would not be running on a web browser and then the problem would not occur.

**Comment by grouma** [comment]
There is an example within the repository. No need for AngularDart. Also there are tests which provide alternative examples.

**Comment by ponderMuse** [comment]
I will have a look at the sse package and see what it does by itself and report back. I should mention however that the current code where I see the issue doesn't use this package at all. The subject name I gave for this issue includes the word SSE but at no point have I ever used [package:sse](https://github.com/dart-lang/sse) itself. The code has so far seen the problem on Chrome web when using [package:http](https://github.com/dart-lang/http) for the client side and also when using [package:eventsource](https://pub.dev/packages/eventsource) for the client side.

**Comment by grouma** [comment]
Wow, big confusion on my part. 

I thought you were having issues with `package:sse` as it is used heavily in the Flutter Web development infrastructure. FYI, Flutter Tool depends on `package:dwds` which depends on `package:sse`.

> Using strictly package:sse and nothing else then the client would not be running on a web browser and then the problem would not occur.

`package:sse` has a backend handler and a corresponding client that runs in an SSE supporting browser such as Chrome.

> The code has so far seen the problem on Chrome web when using package:http for the client side and also when using package:eventsource for the client side.

I'm less familiar with these packages. I'd open up an issue on those packages. Alternatively you can try migrating to `package:sse` as we know this works without issue on Chrome.

**Comment by ponderMuse** [comment]
I have tried the Flutter client using [package:sse](https://github.com/dart-lang/sse) and you are correct in that this package does indeed work in Chrome web. However, the reason it works on Chrome web is because the package uses `dart:html`. But it is my understanding that `dart:html` only works for web apps and therefore it would NOT work if the client were to be run as a mobile app (if my understanding is correct).

The other packages I've tried: [package:eventsource](https://github.com/stevenroose/dart-eventsource) and pure `dart:http` I believe have been updated to remove the need for `dart:html` so that the code can be run in both mobile and web but, it is my guess that, although the code itself compiles and launches in both mobile and web, receiving SSE events does not quite work when running in Chrome web (receiving SSE events only works when running as mobile app).

So, to sum up, from the tests I've run so far:
- [package:eventsource](https://github.com/stevenroose/dart-eventsource) compiles and runs as both mobile and web but SSE events are only received when running as mobile app.
- [package:sse](https://github.com/dart-lang/sse) runs only on web and can therefore only receive SSE events when running as web app.

I have yet to see a working Flutter web client that is able to receive SSE events when running as a mobile app and as a web app. i.e. a Flutter web client that is able to receive SSE events for both mobile and web apps.

**Comment by grouma** [comment]
Thanks for the investigation.

As stated above, I'm not familiar with `package:eventsource`. It's worth opening an issue / reaching out to the author to see if that package was ever intended to work in the browser.

> I have yet to see a working Flutter web client that is able to receive SSE events when running as a mobile app and as a web app.

You can create an abstraction on top of both packages and use conditional imports to use the correct package given the platform. I'm not sure how ergonomic that solution may be. 

Alternatively, I'm open to a pull request on `package:sse` that introduces a new client for use on Flutter mobile. We can again use conditional imports so that it's ergonomic for users of the package regardless of the platform.

**Comment by yjbanov** [comment]
Sounds like there's not much we can do on the Flutter side. I'll close the issue, but feel free to reopen if you believe something in Flutter isn't working.

**Comment by ponderMuse** [comment]
Yes, no problem. This Ticket can be closed here at least.

As @grouma suggested I will for the moment use conditional imports so that when running on a browser it uses `package:sse` and when it is running on mobile it uses `package:eventsource` via an abstracted interface:

	export 'sse_event_clt.dart'
	  if (dart.library.io) 'sse_event_mob.dart'
	  if (dart.library.js) 'sse_event_web.dart';

It would be good however if `package:sse` (and/or any other http client package for that matter) could be updated to work in both browsers and mobiles/tablets out of the box. This would be useful in Flutter Web 1.9+ world for sure. @grouma, is there anything you would like me to do in order to get a new pull request going to get `package:sse` updated to do just this?

**Comment by grouma** [comment]
> It would be good however if package:sse (and/or any other http client package for that matter) could be updated to work in both browsers and mobiles/tablets out of the box. This would be useful in Flutter Web 1.9+ world for sure. @grouma, is there anything you would like me to do in order to get a new pull request going to get package:sse updated to do just this?

I don't know what is required to get `package:sse` working on mobile. I unfortunately won't have time to investigate it either. That being said I'm more than happy to review any PR you send my way.

**Comment by ponderMuse** [comment]
Hello @grouma,

I will pull the sse project down from github when I get a chance and see if I can figure out what it would take to make it work for mobiles/tablets too. Hopefully not too much. But I don't know when I'll get round to it. I am still learning Dart and everything Flutter myself so I don't know if I know enough yet to get the package updated but will sure give it a go. Next time I post I will post in that project as there is no point posting on this thread any more.

Regards,
PM

**Comment by hivesey** [comment]
@ponderMuse I am stuck with the same issue..initially using Dio but that doesn't work on Flutter Web. Do you have any updates to share on this issue?

Thanks.

**Comment by mclabs74** [comment]
my guess is it is because this https://pub.dev/documentation/http/latest/browser_client/BrowserClient-class.html "It is also unable to stream requests or responses; a request will only be sent and a response will only be returned once all the data is available."
